### PR TITLE
Remove deprecated AccumuloVFSClassLoader 

### PR DESCRIPTION
**Pull request by cshannon**
This PR is to resolve issue #3131 as the VFS classloader has been pulled out into an external project https://github.com/apache/accumulo-classloaders . All references to the VFS classloader have been removed (including properties, documentation, tests, etc). The classloader can still be used if desired by configuring the property `GENERAL_CONTEXT_CLASSLOADER_FACTORY` to use the VFS classloader that was pulled into the external project.

### CODE REVIEW

File: start/src/main/java/org/apache/accumulo/start/classloader/vfs/ContextManager.java
Comment (dlmarion) [inline start/src/main/java/org/apache/accumulo/start/classloader/vfs/ContextManager.java:85]: Not sure we need to the parent argument, the parent classloader will be the ClassLoader for this class.

File: core/src/main/java/org/apache/accumulo/core/conf/PropertyType.java
Comment (dlmarion) [inline core/src/main/java/org/apache/accumulo/core/conf/PropertyType.java:111]: I think you can just remove line 113 with no replacement

File: core/src/test/java/org/apache/accumulo/core/conf/AccumuloConfigurationTest.java
Comment (dlmarion) [inline core/src/test/java/org/apache/accumulo/core/conf/AccumuloConfigurationTest.java:270]: Suggest changing these file paths to use file:///

File: core/src/main/java/org/apache/accumulo/core/conf/Property.java
Comment (dlmarion) [inline core/src/main/java/org/apache/accumulo/core/conf/Property.java:1095]: GENERAL_CLASSPATHS property can be removed also.

File: minicluster/src/main/java/org/apache/accumulo/miniclusterImpl/MiniAccumuloConfigImpl.java
Comment (dlmarion) [inline minicluster/src/main/java/org/apache/accumulo/miniclusterImpl/MiniAccumuloConfigImpl.java:156]: This property can be removed also

File: start/src/main/java/org/apache/accumulo/start/Main.java
Comment (dlmarion) [inline start/src/main/java/org/apache/accumulo/start/Main.java:97]: I think this method can be removed.

File: shell/src/main/java/org/apache/accumulo/shell/commands/ClasspathCommand.java
Comment (dlmarion) [inline shell/src/main/java/org/apache/accumulo/shell/commands/ClasspathCommand.java:35]: I'm going to think about how we can still do this.

File: core/src/main/java/org/apache/accumulo/core/conf/Property.java
Comment (ctubbsii) [inline core/src/main/java/org/apache/accumulo/core/conf/Property.java:1083]: This is a new property. I don't think we need this. We are removing this feature in favor of the system classloader, which is already configurable by Java. We never need to specify a classpath in properties, just a context name on a per-table basis, which is interpreted based on the implementation details of the classloader factory, which should be pluggable SPI.

File: shell/src/main/java/org/apache/accumulo/shell/commands/ClasspathCommand.java
Comment (ctubbsii) [reply shell/src/main/java/org/apache/accumulo/shell/commands/ClasspathCommand.java:35]: Showing the contents of the `java.class.path` system property is probably enough here. For a configured context classloader factory, we could extend the SPI to have a method to show information about each context and call that here also. But, since this is run on the client side in the shell, it would have to be set up on the client side, and not just the server-side... unless we had a new RPC to specifically ask a random server to give us that info. And, in any case, we wouldn't be able to have much insight into the user's configured system classloader, if they set that... because that could do anything.

File: start/src/main/java/org/apache/accumulo/start/classloader/vfs/ContextManager.java
Comment (ctubbsii) [inline start/src/main/java/org/apache/accumulo/start/classloader/vfs/ContextManager.java:37]: This class was deprecated with the expectation to remove it entirely. What do we still need it for?

**Review by ctubbsii — CHANGES_REQUESTED** [review]

File: shell/src/main/java/org/apache/accumulo/shell/commands/ClasspathCommand.java
Comment (dlmarion) [reply shell/src/main/java/org/apache/accumulo/shell/commands/ClasspathCommand.java:35]: I think it might be sufficient to remove the classpath command. In the default scenario, the server and client VMs will use the standard JVM classpath mechanism. In the case where a users is using the new VFS classloader, the classpath is [printed](https://github.com/apache/accumulo-classloaders/blob/main/modules/vfs-class-loader/src/main/java/org/apache/accumulo/classloader/vfs/AccumuloVFSClassLoader.java#L596) to the server-side log during initialization (so it's only printed once).

File: core/src/main/java/org/apache/accumulo/core/conf/Property.java
Comment (dlmarion) [reply core/src/main/java/org/apache/accumulo/core/conf/Property.java:1083]: Tables use a classloader name (context) from which to load user iterators and such. The classpaths for those classloaders need to be defined somewhere. It's been a while since we discussed the approach for this when removing VFS. What are your thoughts? I'm thinking that a URLClassLoader would be created per context classpath (it doesn't reload and it only supports what the URLClassLoader supports today).

File: shell/src/main/java/org/apache/accumulo/shell/commands/ClasspathCommand.java
Comment (ctubbsii) [reply shell/src/main/java/org/apache/accumulo/shell/commands/ClasspathCommand.java:35]: Was the command previously deprecated? If not, we could at least dump `java.class.path` to make it *somewhat* useful without just removing it, and possibly causing scripts to break.

File: core/src/main/java/org/apache/accumulo/core/conf/Property.java
Comment (ctubbsii) [reply core/src/main/java/org/apache/accumulo/core/conf/Property.java:1083]: The idea previously discussed was that the `general.context.class.loader.factory` is the singleton service for creating class loaders based on a specified context in the `table.class.loader.context` property. This is also what the descriptions of those properties say. So, the interpretation of the `table.class.loader.context` value is dependent on which implementation factory has been set in `general.context.class.loader.factory`. If you set a URLClassLoader factory, you can just set the `table.class.loader.context` property to be a set of URLs. You don't need a separate property at all. You can also have a more advanced factory that just takes a simplified name, and maps that name to a set of URLs or whatever, in its own configuration file.

Basically, all we need is the factory that does `String context -> ClassLoader` mapping, and the property that stores the `String context`. That context could be a name, a number, a JSON blob, a comma-separated set of URLs (or using another delimiter than a comma), or whatever the user can imagine for their particular implementation.

By default, though, we discussed having a URLClassLoader that interprets the context field as a set of URLs. If you're concerned about misconfiguration, the implementation could require the set of URLs have some kind of prefix to validate, as in `urls:url1,url2,url3,...`, and a separate project could create a VFS classloader that had similar validation, `vfs:url1,url2,url3,...`, so mis-configurations are more detectable. But, validation like that is not a requirement of the implementing factory.

File: shell/src/main/java/org/apache/accumulo/shell/commands/ClasspathCommand.java
Comment (dlmarion) [reply shell/src/main/java/org/apache/accumulo/shell/commands/ClasspathCommand.java:35]: I don't think so, but could be done in 2.1.1.

File: core/src/main/java/org/apache/accumulo/core/conf/Property.java
Comment (dlmarion) [reply core/src/main/java/org/apache/accumulo/core/conf/Property.java:1083]: > By default, though, we discussed having a URLClassLoader that interprets the context field as a set of URLs.

Ok, this is what I was forgetting. I do remember now discussing this. I have no issue with this, a user would set `table.class.loader.context=url1,url2,url3` and `url1,url2,url3` would be the name of the context and the set of URLs for the URLClassLoader.

File: start/src/main/java/org/apache/accumulo/start/classloader/vfs/ContextManager.java
Comment (dlmarion) [reply start/src/main/java/org/apache/accumulo/start/classloader/vfs/ContextManager.java:37]: Given the discussion above about how the DefaultContextClassLoaderFactory will work now, I think this can be removed.

**Comment by cshannon** [comment]
Thanks for the feedback @ctubbsii  and @dlmarion , I will rework this to remove the new property I added and to just have the existing properties and let the user provide the implementation for the class loader factory. I'll be looking at this again right after the new year to make the updates.

**Comment by cshannon** [comment]
I started taking a peak at this again and I should hopefully have time to update with the changes requested this upcoming week at some point.

**Comment by cshannon** [comment]
@ctubbsii and @dlmarion - I just pushed a new update and I have significantly reworked the PR. There's still work to be done...comments need updating, some tests still need fixing and to be updated, and there is still some old VFS references laying around, etc. But it's in a good enough state to be looked at again to make sure this new direction is correct before I finish polishing everything up.

The latest update now implements the agreed upon behavior of just re-using the existing 2 properties `general.context.class.loader.factory` and `table.class.loader.context`

If not specified the `DefaultContextClassLoaderFactory` now will just load a new `URLClassLoader` and use the table context property as a list of URIs. I got rid of the ContextManager and simplified things to just cache classloaders for a period of time in a cache (for how long is something I wasn't sure of yet). I got rid of the complex logic to remove contexts as it seemed unnecessary and a simple Caffeine cache should work.

File: core/src/main/java/org/apache/accumulo/core/conf/Property.java
Comment (dlmarion) [inline core/src/main/java/org/apache/accumulo/core/conf/Property.java:1073]: I think we need a property like this, maybe called GENERAL_CONTEXT_CLASSPATH_PREFIX, so that administrators/users can define the classpaths for a context. Unless you accounted for this somewhere else, it's possible I missed it. The current way users interact with this is to define named classloader context and their paths, for example:

```
general.vfs.context.classpath.<contextName>=<some classpath>
```

Then, in supported places in the code, they can pass in the <contextName> to use their custom supplied classes.

Comment (dlmarion) [reply core/src/main/java/org/apache/accumulo/core/conf/Property.java:1073]: @cshannon reminded me that this is not needed based on earlier discussions. The context property set on the table will contain a comma separated list of URIs.

File: shell/src/main/java/org/apache/accumulo/shell/commands/ClasspathCommand.java
Comment (dlmarion) [reply shell/src/main/java/org/apache/accumulo/shell/commands/ClasspathCommand.java:35]: So, I think that it could print `java.class.path` (of the shell, not the server, unless they are the same) and then print out all of the values of the `table.class.loader.context` properties defined on the tables.

File: start/src/main/java/org/apache/accumulo/start/Main.java
Comment (dlmarion) [inline start/src/main/java/org/apache/accumulo/start/Main.java:46]: Do we need the `getClassLoader()` method anymore? Can we just call `ClassLoaderUtil.getClassLoader(null);` ?

**Review by dlmarion — COMMENTED** [review]
I didn't look at the tests yet. My thought on that is that only the tests that were testing the table context need to remain, the tests that were testing the VFSClassLoader as the system classloader can probably be removed.

File: core/src/main/java/org/apache/accumulo/core/classloader/DefaultContextClassLoaderFactory.java
Comment (cshannon) [reply core/src/main/java/org/apache/accumulo/core/classloader/DefaultContextClassLoaderFactory.java:60]: I agree, we should probably change the name to something else. The value can actually be anything you want. By default it would be a list of URIs because that's what the default factory would use to create a classloader but if a user plugs in their own class loader factory implementation then this value can be anything the user wants as long as their factory knows how to read it. So I think `contextValue` or something like that makes a lot more sense as `contextName` doesn't apply anymore.

File: start/src/main/java/org/apache/accumulo/start/Main.java
Comment (cshannon) [reply start/src/main/java/org/apache/accumulo/start/Main.java:46]: I actually tried to do that but the issue is ClassLoaderUtil isn't available to use here. The main class is part of the `start` module and ClassLoaderUtil lives inside the `core` module. Core has a dependency on start so we can't use the class.

File: core/src/main/java/org/apache/accumulo/core/classloader/DefaultContextClassLoaderFactory.java
Comment (cshannon) [reply core/src/main/java/org/apache/accumulo/core/classloader/DefaultContextClassLoaderFactory.java:71]: The purpose of Caffeine was so that the cache will remove old classloaders automatically when no longer needed but this could be re-worked if desired. 

In the current version there is a thread that runs to check for no longer needed contexts and will remove them. The idea of using Caffeine was that we can cache the Classloader but if it is no longer needed anymore after some period of time (nothing reads it because maybe the config was dropped, etc) we can then just remove it from the cache automatically. This seemed simpler than maintaining a GC thread.

Comment (dlmarion) [reply core/src/main/java/org/apache/accumulo/core/classloader/DefaultContextClassLoaderFactory.java:71]: I understand the need for Caffeine, I'm just not sure we need this class. We could change the definition of the Caffeine map from <String,Context> to <String,Classloader>.

Comment (cshannon) [reply core/src/main/java/org/apache/accumulo/core/classloader/DefaultContextClassLoaderFactory.java:71]: Ah yeah sorry, I misread what you were saying. The Context class could be removed and I'll do that, it was a hold over from before. I had kept it in case for some reason we wanted to add more metadata to what is cached but storing the ClassLoader as the value should be fine as I can't really think of a good reason we need to have anything else cached and it could always be changed later.

File: core/src/main/java/org/apache/accumulo/core/classloader/DefaultContextClassLoaderFactory.java
Comment (ctubbsii) [reply core/src/main/java/org/apache/accumulo/core/classloader/DefaultContextClassLoaderFactory.java:60]: I was just calling it "context". The semantics of the context, whether it's a URL, a path, a name, etc., depends on the implementation. But generically, it's just a context, and the factory's job is to map a context to a ClassLoader, effectively equivalent to `Function<String,ClassLoader>`. You could call it `context` or `contextString` or similar. I agree it doesn't make sense to call it a `name`, though it could be a name for some implementations.

Another thing about naming: this should not be called `DefaultContextClassLoaderFactory`. It should be named for how it behaves... not it's current status as the default. It will be very confusing (and I've seen this happen) if the default is changed, and it's no longer the default. So, for example, the default could change from `DefaultThing` to `MyThing`, and so then when people say `default thing`, you don't know if they mean the default `MyThing` or the non-default `DefaultThing`. In this case, it could just be `URLClassLoaderFactory` or `URLContextClassLoaderFactory`, and return instances of `URLClassLoader` objects, given a context that represents a list of URLs.

**Comment by cshannon** [comment]
@ctubbsii and @dlmarion - This should be ready for a final review, I ran through the sunny IT tests and they passed. I also have kicked off a full IT build as well.

File: start/src/main/java/org/apache/accumulo/start/Main.java
Comment (ctubbsii) [inline start/src/main/java/org/apache/accumulo/start/Main.java:50]: Pretty sure this stuff can now be removed. It was added for a very specific workaround that I don't think applies anymore: https://issues.apache.org/jira/browse/ACCUMULO-4341

File: shell/src/main/java/org/apache/accumulo/shell/commands/ClasspathCommand.java
Comment (ctubbsii) [inline shell/src/main/java/org/apache/accumulo/shell/commands/ClasspathCommand.java:51]:
```suggestion
      writer.println("Accumulo Shell Classpath:");
```

File: shell/src/main/java/org/apache/accumulo/shell/commands/ClasspathCommand.java
Comment (ctubbsii) [inline shell/src/main/java/org/apache/accumulo/shell/commands/ClasspathCommand.java:59]:
```suggestion
      Arrays.stream(javaClassPath.split(File.pathSeparator)).forEach(classPathUri -> writer.println(classPathUri));
```

File: shell/src/main/java/org/apache/accumulo/shell/commands/ClasspathCommand.java
Comment (ctubbsii) [inline shell/src/main/java/org/apache/accumulo/shell/commands/ClasspathCommand.java:61]:
```suggestion
      writer.println();
```

File: shell/src/main/java/org/apache/accumulo/shell/commands/ClasspathCommand.java
Comment (ctubbsii) [inline shell/src/main/java/org/apache/accumulo/shell/commands/ClasspathCommand.java:62]: I would just catch the checked exceptions here using multi-catch syntax, so you don't rewrap RTEs.

File: shell/src/main/java/org/apache/accumulo/shell/commands/ClasspathCommand.java
Comment (ctubbsii) [inline shell/src/main/java/org/apache/accumulo/shell/commands/ClasspathCommand.java:63]: Is there a more specific RTE type that we can use instead of this?

File: core/src/main/java/org/apache/accumulo/core/file/blockfile/cache/impl/BlockCacheManagerFactory.java
Comment (ctubbsii) [inline core/src/main/java/org/apache/accumulo/core/file/blockfile/cache/impl/BlockCacheManagerFactory.java:34]: I think this will now use the system class loader. I don't think this is done inside a context, but I am not sure.

**Review by ctubbsii — CHANGES_REQUESTED** [review]

**Comment by ctubbsii** [comment]
I also think it would be good to compare the before/after on the shell's classpath command and the bin/accumulo classpath command to make sure we didn't lose anything important in the output options.

**Comment by cshannon** [comment]
Something else I realized is I forgot to test this against https://github.com/apache/accumulo-classloaders to verify we can still use the VFS classloader using this new approach. That's something I will do as well (assuming the full IT tests pass and everything is good)

File: start/src/main/java/org/apache/accumulo/start/Main.java
Comment (cshannon) [reply start/src/main/java/org/apache/accumulo/start/Main.java:50]: I can remove this but I was thinking that wouldn't this issue still be a problem if a user configured Accumulo to use the VFS classloader that is now external?

Comment (dlmarion) [reply start/src/main/java/org/apache/accumulo/start/Main.java:50]: Going back and looking at ACCUMULO-4341 and ACCUMULO-3923, it looks like this was added to get around an issue, that's been resolved, in VFS not loading resource files in jars correctly. I think this can be removed. If you test this with the new VFS classloader and we run into an issue, then we can put it back.

**Review by dlmarion — COMMENTED** [review]
Looks good to me, one comment and one question

**Comment by cshannon** [comment]
I ran the full ITs and noticed just 1 failure in `org.apache.accumulo.test.functional.ClassLoaderIT` which I need to investigate but the other tests all passed.

**Review by dlmarion — APPROVED** [review]

**Comment by cshannon** [comment]
I think this is ready to go, I pushed more updates including removing the broken ClassLoaderIT that was failing as that no loner applies (it only was there to test the VFS reloading) and I also removed the no longer needed code from the Main start class that also only applied to VFS. 

I spent a bunch of time testing to make sure the VFS classloader worked with the changes by using the [accumulo-classloaders](https://github.com/apache/accumulo-classloaders) project and seems to be working well. I used these instructions https://github.com/apache/accumulo-classloaders/blob/main/modules/vfs-class-loader/TESTING.md and I tested against Uno. I tested both Setting a system classloader as well as using a context using the new [instructions](https://github.com/apache/accumulo-classloaders/blob/main/modules/vfs-class-loader/TESTING.md#setting-scan-context-on-table-new) and properties.

A couple changes will be needed to the documentation to the instructions page (property rename for the table context, need to mention to make sure to add the commons VFS jar since it was removed from 3.0.0) otherwise the testing instructions are pretty much up to date. I can submit a small PR with the changes probably tomorrow.

I kicked off another full IT so this should be ready to merge if that passes.

**Comment by cshannon** [comment]
Full IT build passed all the tests for this, I also just pushed a PR to update testing instructions for accumulo-classloaders: https://github.com/apache/accumulo-classloaders/pull/17

@ctubbsii - I think this is ready to merge, did you want to do a final review?

**Comment by ctubbsii** [comment]
> @ctubbsii - I think this is ready to merge, did you want to do a final review?

You merged while I was giving it another lookover, but I don't see anything that stands out. I think you addressed all the previous things I noticed.

**Comment by cshannon** [comment]
> > @ctubbsii - I think this is ready to merge, did you want to do a final review?
> 
> You merged while I was giving it another lookover, but I don't see anything that stands out. I think you addressed all the previous things I noticed.

Sorry about that, it had been about a week so figured it was good to go since @dlmarion said it looked good. I also figured if we found something after that needed fixing I could go back and fix it with a follow on PR.

**Comment by ctubbsii** [comment]
No worries. You did the right thing. I was merely observing the coincidence. That we both were looking at it in the same few minutes after a week of inaction was amusing to me 😁

**Pull request merged**
